File: threedollar-api-user-service/src/main/java/com/depromeet/threedollar/api/userservice/listener/slack/SlackNotificationEventListener.java
package com.depromeet.threedollar.api.userservice.listener.slack;

import static com.depromeet.threedollar.infrastructure.external.client.slack.type.SlackNotificationMessageType.ERROR_MESSAGE;
import static com.depromeet.threedollar.infrastructure.external.client.slack.type.SlackNotificationMessageType.INFO_MESSAGE;

import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import com.depromeet.threedollar.common.model.event.ApplicationStateChangedEvent;
import com.depromeet.threedollar.common.model.event.ServerExceptionOccurredEvent;
import com.depromeet.threedollar.infrastructure.external.client.slack.SlackWebhookApiClient;
import com.depromeet.threedollar.infrastructure.external.client.slack.dto.request.PostSlackMessageRequest;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Component
public class SlackNotificationEventListener {

    private final SlackWebhookApiClient slackNotificationApiClient;

    @Async
    @EventListener
    public void sendErrorNotification(ServerExceptionOccurredEvent event) {
        slackNotificationApiClient.postMonitoringMessage(PostSlackMessageRequest.of(ERROR_MESSAGE.generateMessage(
            event.getApplicationType().getDescription(),
            event.getErrorCode().getCode(),
            event.getRequestUri(),
            event.getException(),
            event.getTimeStamp(),
            event.getErrorCode().getMessage(),
            event.getUserMetaValue()
        )));
    }

    @Async
    @EventListener
    public void sendInfoNotification(ApplicationStateChangedEvent event) {
        slackNotificationApiClient.postMonitoringMessage(PostSlackMessageRequest.of(INFO_MESSAGE.generateMessage(
            event.getApplicationType().getDescription(),
            event.getApplicationUid(),
            event.getMessage(),
            event.getTimeStamp()
        )));
    }

}

File: threedollar-api-admin-service/src/main/kotlin/com/depromeet/threedollar/api/adminservice/listener/slack/SlackNotificationEventListener.kt
package com.depromeet.threedollar.api.adminservice.listener.slack

import com.depromeet.threedollar.common.model.event.ApplicationStateChangedEvent
import com.depromeet.threedollar.infrastructure.external.client.slack.SlackWebhookApiClient
import com.depromeet.threedollar.infrastructure.external.client.slack.dto.request.PostSlackMessageRequest
import com.depromeet.threedollar.infrastructure.external.client.slack.type.SlackNotificationMessageType
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

@Component
class SlackNotificationEventListener(
    private val slackNotificationApiClient: SlackWebhookApiClient,
) {

    @EventListener
    fun sendInfoNotification(event: ApplicationStateChangedEvent) {
        slackNotificationApiClient.postMonitoringMessage(
            PostSlackMessageRequest.of(
                SlackNotificationMessageType.INFO_MESSAGE.generateMessage(
                    event.applicationType.description,
                    event.applicationUid,
                    event.message,
                    event.timeStamp
                )))
    }

}
